### PR TITLE
Compilation of functors

### PR DESCRIPTION
**Pull request by Ekdohibs**
This PR implements the compilation of functors, through defunctorization. Most notably, functor application is always generative, so not a lot of constructions are supported, but this should be enough for `{Set,Map}.Make`.

### CODE REVIEW

File: miniml/compiler/compile.scm
Comment (gasche) [inline miniml/compiler/compile.scm:1780]: If you use pattern-matching instead of an assertion+cadr here,

```
(match-let* ((('VFunctor fct) mod)
             (nenv ...))
```

you get the following side-benefits;
- (minor) a trace in the code of the expected number of arguments of the `'Vfunctor` constructor
- (major) if an input does not fit here, you get a proper match-failure error from guile that shows the offending input, instead of a less-informative assertion failure; this has really helped me debug compiler code

File: miniml/compiler/compile.scm
Comment (gasche) [inline miniml/compiler/compile.scm:974]: same here, pattern-matching with `match-let` would result in more readable failure messages than `assert`

File: miniml/compiler/compile.scm
Comment (gasche) [inline miniml/compiler/compile.scm:104]: Using the prefix `M` is a bit confusing here, as it is used above for structure items. Maybe `ME`?

Comment (gasche) [reply miniml/compiler/compile.scm:104]: I think the code would be better if you added a `FUNCTOR functor_arg MINUSGT module_expr` case here, and forced yourself to handle cleanly it below.

File: miniml/compiler/compile.scm
Comment (gasche) [inline miniml/compiler/compile.scm:1831]: Given the way it is used, `compile-functor` should be called `compile-module` I think.

File: miniml/compiler/compile.scm
Comment (gasche) [inline miniml/compiler/compile.scm:1790]: There is a lot of redundancy between the end of this function and the one above.

Comment (gasche) [reply miniml/compiler/compile.scm:1790]: Could we do without the `F` wrapping? If you had a `MEFunctor` as suggested above, you could just use this to build the `fct` module expression (could be called `mod-expr`). Then you call `compile-mod` on it in any case, and that function handles `MEFunctor` by returning a `VFunctor`. How does that sound?

**Comment by gasche** [comment]
> functor application is always generative, so not a lot of constructions are supported

For me applicative vs. generative is mostly a question of whether (generative) datatype definitions are distinct or compatible in two applications. Is there a difference here? My understanding would be that, as we are only using datatypes for representation and not for type-checking, two instances of the same datatype will in fact be interchangeable in the compiled program, so in this sense we are not non-applicative. What kind of incompatibilities do you have in mind?

File: miniml/compiler/compile.scm
Comment (gasche) [reply miniml/compiler/compile.scm:104]: Note: there is no strong need to add a production to anonymous functors of course, just using a `MEFunctor` constructor below simplifies the picture.

**Comment by Ekdohibs** [comment]
I've changed the code to add `'MEFunctor`, and the result is indeed quite a bit simpler, thanks!
Concerning applicative vs. generative, it is the same concerning datatypes, yes. The incompatibilities I have in mind would happen in the case of a functor like `module X(A : sig val x : int end) = struct let r = ref 0 end` for instance, where instantiating the functor has side-effects.

**Comment by gasche** [comment]
I'm not sure which differents you have in mind; I think that dynamically uses of `F(A)` that are not within types/signatures do result in evaluation in OCaml, even when `F` is applicative, so they should be the same as in your proposal.

File: miniml/compiler/hello.ml
Comment (gasche) [inline miniml/compiler/hello.ml:280]: Nitpick: I think it could be useful to have a test where we call `F(X)` several times, checking that while the effects in `F` are duplicated, the effects in the evaluation of `X` (previously defined as `module X = struct ... end`) are not, that its components are reduced to values and then shared between functor applications.

Comment (gasche) [reply miniml/compiler/hello.ml:280]: (Also nitpick: I am more familiar with naming conventions where functors are called `F, G`, functor arguments are `M` or `X, Y, Z`, and modules are something like `M` or `A, B, C`.)

**Review by gasche — APPROVED** [review]
This looks good to me, thanks. I wonder if you want to do other changes before merging, and whether you expect to merge yourself; otherwise I would be happy to merge.

**Comment by Ekdohibs** [comment]
I don't mind merging myself, but would be very happy if you did :). I agree about checking the effects of the argument are not duplicated should be done.

**Comment by gasche** [comment]
Then I'll wait for the testsuite update (I guess?) and merge.

**Comment by Ekdohibs** [comment]
Done!

**Comment by gasche** [comment]
Thanks, merged.

**Pull request merged**
